File: pool/coinbase.go
package pool

import (
	"encoding/binary"
	"encoding/hex"
	"math/big"
	"strings"

	"zpoolproxy/config"
)

type CoinbaseParts struct {
	Coinb1 string
	Coinb2 string
}

// EncodeBip34Height encodes block height into BIP34 push script bytes
func EncodeBip34Height(height int64) []byte {
	if height == 0 {
		return []byte{0x01, 0x00}
	}
	var bytes []byte
	temp := height
	for temp > 0 {
		bytes = append(bytes, byte(temp&0xff))
		temp >>= 8
	}
	if bytes[len(bytes)-1]&0x80 != 0 {
		bytes = append(bytes, 0x00)
	}
	return append([]byte{byte(len(bytes))}, bytes...)
}

const base58Alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"

func base58Decode(input string) []byte {
	radix := big.NewInt(58)
	result := big.NewInt(0)

	for i := 0; i < len(input); i++ {
		char := input[i]
		idx := strings.IndexByte(base58Alphabet, char)
		if idx == -1 {
			return nil
		}
		result.Mul(result, radix)
		result.Add(result, big.NewInt(int64(idx)))
	}

	b := result.Bytes()

	// Count leading '1's
	var numZeros int
	for i := 0; i < len(input) && input[i] == '1'; i++ {
		numZeros++
	}

	res := make([]byte, numZeros+len(b))
	copy(res[numZeros:], b)
	return res
}

func bech32DecodeToScriptPubKey(address string) []byte {
	clean := strings.ToLower(strings.TrimSpace(address))
	parts := strings.Split(clean, "1")
	if len(parts) != 2 {
		return nil
	}
	dataPart := parts[1]
	if len(dataPart) < 6 {
		return nil
	}

	const charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
	var values []byte
	for i := 0; i < len(dataPart)-6; i++ {
		idx := strings.IndexByte(charset, dataPart[i])
		if idx == -1 {
			return nil
		}
		values = append(values, byte(idx))
	}

	if len(values) == 0 {
		return nil
	}

	witnessVersion := values[0]
	program5bit := values[1:]

	// Convert 5-bit array to 8-bit buffer
	acc := 0
	bits := 0
	var program8bit []byte
	for _, val := range program5bit {
		acc = (acc << 5) | int(val)
		bits += 5
		for bits >= 8 {
			bits -= 8
			program8bit = append(program8bit, byte((acc>>bits)&0xff))
		}
	}

	var witnessOp byte
	if witnessVersion == 0 {
		witnessOp = 0x00
	} else {
		witnessOp = 0x50 + witnessVersion
	}

	var script []byte
	script = append(script, witnessOp, byte(len(program8bit)))
	script = append(script, program8bit...)
	return script
}

// AddressToScriptPubKey converts P2PKH/P2SH/Bech32 address or raw Hex script to hex scriptPubKey bytes
func AddressToScriptPubKey(address string) []byte {
	clean := strings.TrimSpace(address)
	fallback, _ := hex.DecodeString("76a914000000000000000000000000000000000000000088ac")

	if clean == "" {
		return fallback
	}

	// 0. Raw Hex ScriptPubKey (e.g. 76a914...88ac or a914...87 or 0014...)
	if rawBytes, err := hex.DecodeString(clean); err == nil && len(rawBytes) >= 10 {
		return rawBytes
	}

	// 1. Bech32 / Segwit (bc1... / tb1... / bcrt1... / dgb1... / ltc1... / any string containing '1')
	if strings.Contains(clean, "1") {
		script := bech32DecodeToScriptPubKey(clean)
		if script != nil {
			return script
		}
	}

	// 2. Base58 (P2PKH / P2SH for any SHA-256 altcoin)
	decoded := base58Decode(clean)
	if len(decoded) >= 25 {
		payload := decoded[len(decoded)-24 : len(decoded)-4] // Extract 20-byte hash payload
		versionByte := decoded[0]

		// P2SH (version 0x05 / 0xc4 / 0x3f)
		if versionByte == 0x05 || versionByte == 0xc4 || versionByte == 0x3f {
			var script []byte
			prefix, _ := hex.DecodeString("a914")
			suffix, _ := hex.DecodeString("87")
			script = append(script, prefix...)
			script = append(script, payload...)
			script = append(script, suffix...)
			return script
		}

		// P2PKH (Universal for Bitcoin '1', DigiByte 'A'/'D', Bitcoin Cash, BSV, Luckycoin, Pepecoin, Litecoin, Testnet)
		var script []byte
		prefix, _ := hex.DecodeString("76a914")
		suffix, _ := hex.DecodeString("88ac")
		script = append(script, prefix...)
		script = append(script, payload...)
		script = append(script, suffix...)
		return script
	}

	return fallback
}

// BuildCoinbaseTransaction constructs the coinb1 and coinb2 hex strings
func BuildCoinbaseTransaction(
	cfg *config.Config,
	blockHeight int64,
	coinbaseValue int64,
	minerAddress string,
	extranonce1Size int,
	extranonce2Size int,
	defaultWitnessCommitment string,
) CoinbaseParts {
	versionBuf := make([]byte, 4)
	binary.LittleEndian.PutUint32(versionBuf, 1)

	inputCountBuf := []byte{0x01}
	prevTxIdBuf := make([]byte, 32)
	prevOutIdxBuf, _ := hex.DecodeString("ffffffff")

	heightPushBuf := EncodeBip34Height(blockHeight)
	totalExtranonceSize := extranonce1Size + extranonce2Size
	scriptLen := len(heightPushBuf) + totalExtranonceSize
	scriptLenBuf := []byte{byte(scriptLen)}
	sequenceBuf, _ := hex.DecodeString("ffffffff")

	// COINB1
	var coinb1Buf []byte
	coinb1Buf = append(coinb1Buf, versionBuf...)
	coinb1Buf = append(coinb1Buf, inputCountBuf...)
	coinb1Buf = append(coinb1Buf, prevTxIdBuf...)
	coinb1Buf = append(coinb1Buf, prevOutIdxBuf...)
	coinb1Buf = append(coinb1Buf, scriptLenBuf...)
	coinb1Buf = append(coinb1Buf, heightPushBuf...)

	// OUTPUTS
	var feeSatoshis int64 = 0
	if cfg.PoolFeePercent > 0 && cfg.PoolFeeAddress != "" {
		feeSatoshis = int64(float64(coinbaseValue) * (cfg.PoolFeePercent / 100.0))
	}
	minerSatoshis := coinbaseValue - feeSatoshis

	scriptPubKey := AddressToScriptPubKey(minerAddress)

	minerValueBuf := make([]byte, 8)
	binary.LittleEndian.PutUint64(minerValueBuf, uint64(minerSatoshis))
	scriptPubKeyLenBuf := []byte{byte(len(scriptPubKey))}

	var minerOutputBuf []byte
	minerOutputBuf = append(minerOutputBuf, minerValueBuf...)
	minerOutputBuf = append(minerOutputBuf, scriptPubKeyLenBuf...)
	minerOutputBuf = append(minerOutputBuf, scriptPubKey...)

	var feeOutputBuf []byte
	hasFeeOutput := false
	if feeSatoshis > 0 && cfg.PoolFeeAddress != "" {
		feeScriptPubKey := AddressToScriptPubKey(cfg.PoolFeeAddress)
		feeValueBuf := make([]byte, 8)
		binary.LittleEndian.PutUint64(feeValueBuf, uint64(feeSatoshis))
		feeScriptLenBuf := []byte{byte(len(feeScriptPubKey))}

		feeOutputBuf = append(feeOutputBuf, feeValueBuf...)
		feeOutputBuf = append(feeOutputBuf, feeScriptLenBuf...)
		feeOutputBuf = append(feeOutputBuf, feeScriptPubKey...)
		hasFeeOutput = true
	}

	var witnessOutputBuf []byte
	if defaultWitnessCommitment != "" {
		witnessValueBuf := make([]byte, 8)
		commitmentBuf, err := hex.DecodeString(defaultWitnessCommitment)
		if err == nil {
			commitmentLenBuf := []byte{byte(len(commitmentBuf))}
			witnessOutputBuf = append(witnessOutputBuf, witnessValueBuf...)
			witnessOutputBuf = append(witnessOutputBuf, commitmentLenBuf...)
			witnessOutputBuf = append(witnessOutputBuf, commitmentBuf...)
		}
	}

	totalOutputs := 1
	if hasFeeOutput {
		totalOutputs++
	}
	if len(witnessOutputBuf) > 0 {
		totalOutputs++
	}
	outputCountBuf := []byte{byte(totalOutputs)}
	lockTimeBuf := make([]byte, 4)

	// COINB2
	var coinb2Buf []byte
	coinb2Buf = append(coinb2Buf, sequenceBuf...)
	coinb2Buf = append(coinb2Buf, outputCountBuf...)
	coinb2Buf = append(coinb2Buf, minerOutputBuf...)
	coinb2Buf = append(coinb2Buf, feeOutputBuf...)
	coinb2Buf = append(coinb2Buf, witnessOutputBuf...)
	coinb2Buf = append(coinb2Buf, lockTimeBuf...)

	return CoinbaseParts{
		Coinb1: hex.EncodeToString(coinb1Buf),
		Coinb2: hex.EncodeToString(coinb2Buf),
	}
}

File: src/pool/coinbase.ts
import crypto from 'crypto';
import { config } from '../config';

/**
 * Convert Bitcoin Address (P2PKH, P2SH, Bech32 Segwit) to Hex scriptPubKey
 */
export function addressToScriptPubKey(address: string): Buffer {
  const cleanAddr = address.trim();

  // 1. Bech32 / Bech32m Segwit (bc1... / tb1... / bcrt1...)
  if (cleanAddr.startsWith('bc1') || cleanAddr.startsWith('tb1') || cleanAddr.startsWith('bcrt1')) {
    return bech32ToScriptPubKey(cleanAddr);
  }

  // 2. Base58 (P2PKH / P2SH)
  const decoded = base58Decode(cleanAddr);
  if (!decoded || decoded.length < 25) {
    throw new Error(`Invalid Base58 Bitcoin Address: ${address}`);
  }

  const payload = decoded.subarray(1, 21); // 20-byte hash
  const versionByte = decoded[0];

  // P2PKH (Mainnet 0x00, Testnet/Regtest 0x6f) -> OP_DUP OP_HASH160 <20> OP_EQUALVERIFY OP_CHECKSIG
  if (versionByte === 0x00 || versionByte === 0x6f) {
    return Buffer.concat([
      Buffer.from('76a914', 'hex'),
      payload,
      Buffer.from('88ac', 'hex'),
    ]);
  }

  // P2SH (Mainnet 0x05, Testnet/Regtest 0xc4) -> OP_HASH160 <20> OP_EQUAL
  if (versionByte === 0x05 || versionByte === 0xc4) {
    return Buffer.concat([
      Buffer.from('a914', 'hex'),
      payload,
      Buffer.from('87', 'hex'),
    ]);
  }

  // Fallback P2PKH
  return Buffer.concat([
    Buffer.from('76a914', 'hex'),
    payload,
    Buffer.from('88ac', 'hex'),
  ]);
}

/**
 * Base58 decoding implementation
 */
const ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';
const BASE58_MAP: Record<string, number> = {};
for (let i = 0; i < ALPHABET.length; i++) {
  BASE58_MAP[ALPHABET.charAt(i)] = i;
}

function base58Decode(str: string): Buffer | null {
  if (str.length === 0) return null;
  let bytes = [0];
  for (let i = 0; i < str.length; i++) {
    const char = str[i];
    const value = BASE58_MAP[char];
    if (value === undefined) return null;

    let carry = value;
    for (let j = 0; j < bytes.length; j++) {
      carry += bytes[j] * 58;
      bytes[j] = carry & 0xff;
      carry >>= 8;
    }

    while (carry > 0) {
      bytes.push(carry & 0xff);
      carry >>= 8;
    }
  }

  for (let i = 0; i < str.length && str[i] === '1'; i++) {
    bytes.push(0);
  }

  return Buffer.from(bytes.reverse());
}

/**
 * Basic Bech32 to scriptPubKey converter
 */
function bech32ToScriptPubKey(address: string): Buffer {
  const parts = address.toLowerCase().split('1');
  if (parts.length !== 2) throw new Error('Invalid Bech32 Address format');
  const dataPart = parts[1];

  const CHARSET = 'qpzry9x8gf2tvdw0s3jn54khce6mua7l';
  const decodedBytes: number[] = [];
  for (let i = 0; i < dataPart.length - 6; i++) {
    const idx = CHARSET.indexOf(dataPart[i]);
    if (idx === -1) throw new Error('Invalid Bech32 character');
    decodedBytes.push(idx);
  }

  const witnessVersion = decodedBytes[0];
  const program5bit = decodedBytes.slice(1);
  
  // Convert 5-bit array to 8-bit buffer
  let acc = 0;
  let bits = 0;
  const program8bit: number[] = [];
  for (const value of program5bit) {
    acc = (acc << 5) | value;
    bits += 5;
    while (bits >= 8) {
      bits -= 8;
      program8bit.push((acc >> bits) & 0xff);
    }
  }

  const programBuf = Buffer.from(program8bit);
  // OP_0 (0x00) + PushLength + Program
  const witnessOp = witnessVersion === 0 ? 0x00 : 0x50 + witnessVersion;
  return Buffer.concat([
    Buffer.from([witnessOp, programBuf.length]),
    programBuf,
  ]);
}

/**
 * Encode BIP34 Block Height into script push bytes
 */
function encodeBip34Height(height: number): Buffer {
  if (height === 0) return Buffer.from([0x01, 0x00]);
  const bytes: number[] = [];
  let temp = height;
  while (temp > 0) {
    bytes.push(temp & 0xff);
    temp >>= 8;
  }
  // Add positive sign byte if top bit is set
  if (bytes[bytes.length - 1] & 0x80) {
    bytes.push(0x00);
  }
  return Buffer.concat([Buffer.from([bytes.length]), Buffer.from(bytes)]);
}

export interface CoinbaseParts {
  coinb1: string;
  coinb2: string;
}

/**
 * Create CKPool Solo Coinbase Transaction split into coinb1 & coinb2
 */
export function buildCoinbaseTransaction(params: {
  blockHeight: number;
  coinbaseValue: number;
  minerAddress: string;
  extranonce1Size: number;
  extranonce2Size: number;
  defaultWitnessCommitment?: string;
}): CoinbaseParts {
  const versionBuf = Buffer.alloc(4);
  versionBuf.writeUInt32LE(1, 0); // Version 1

  const inputCountBuf = Buffer.from([0x01]); // 1 input

  const prevTxIdBuf = Buffer.alloc(32, 0); // 32 bytes of zeros
  const prevOutIdxBuf = Buffer.from('ffffffff', 'hex'); // 0xffffffff

  // Height push
  const heightPushBuf = encodeBip34Height(params.blockHeight);

  // Pool Signature text
  const poolTextBuf = Buffer.from(config.coinbaseText, 'utf8');

  // Part of coinbase script BEFORE extranonce1
  const scriptPrefix = Buffer.concat([heightPushBuf, poolTextBuf]);

  // Total extranonce length = extranonce1 + extranonce2
  const totalExtranonceSize = params.extranonce1Size + params.extranonce2Size;

  // Total script length = scriptPrefix length + totalExtranonceSize
  const scriptLen = scriptPrefix.length + totalExtranonceSize;
  const scriptLenBuf = Buffer.from([scriptLen]);

  const sequenceBuf = Buffer.from('ffffffff', 'hex');

  // COINB1 = Everything from Tx Version down to end of scriptPrefix
  const coinb1Buf = Buffer.concat([
    versionBuf,
    inputCountBuf,
    prevTxIdBuf,
    prevOutIdxBuf,
    scriptLenBuf,
    scriptPrefix,
  ]);

  // OUTPUTS
  let feeSatoshis = 0;
  if (config.poolFeePercent > 0 && config.poolFeeAddress) {
    feeSatoshis = Math.floor(params.coinbaseValue * (config.poolFeePercent / 100));
  }
  const minerSatoshis = params.coinbaseValue - feeSatoshis;

  let scriptPubKey: Buffer;
  try {
    scriptPubKey = addressToScriptPubKey(params.minerAddress);
  } catch (err) {
    // Fallback unspendable script if invalid address provided
    scriptPubKey = Buffer.from('76a914000000000000000000000000000000000000000088ac', 'hex');
  }

  const minerValueBuf = Buffer.alloc(8);
  minerValueBuf.writeBigUInt64LE(BigInt(minerSatoshis), 0);
  const scriptPubKeyLenBuf = Buffer.from([scriptPubKey.length]);
  const minerOutputBuf = Buffer.concat([minerValueBuf, scriptPubKeyLenBuf, scriptPubKey]);

  let feeOutputBuf = Buffer.alloc(0);
  let hasFeeOutput = false;
  if (feeSatoshis > 0 && config.poolFeeAddress) {
    try {
      const feeScriptPubKey = addressToScriptPubKey(config.poolFeeAddress);
      const feeValueBuf = Buffer.alloc(8);
      feeValueBuf.writeBigUInt64LE(BigInt(feeSatoshis), 0);
      const feeScriptLenBuf = Buffer.from([feeScriptPubKey.length]);
      feeOutputBuf = Buffer.concat([feeValueBuf, feeScriptLenBuf, feeScriptPubKey]);
      hasFeeOutput = true;
    } catch (e) {
      console.error('[Coinbase] Invalid POOL_FEE_ADDRESS in config, skipping fee output');
    }
  }

  let witnessOutputBuf = Buffer.alloc(0);
  if (params.defaultWitnessCommitment) {
    const witnessValueBuf = Buffer.alloc(8, 0); // 0 satoshis
    const commitmentBuf = Buffer.from(params.defaultWitnessCommitment, 'hex');
    const commitmentLenBuf = Buffer.from([commitmentBuf.length]);
    witnessOutputBuf = Buffer.concat([witnessValueBuf, commitmentLenBuf, commitmentBuf]);
  }

  let totalOutputs = 1;
  if (hasFeeOutput) totalOutputs++;
  if (params.defaultWitnessCommitment) totalOutputs++;
  const outputCountBuf = Buffer.from([totalOutputs]);

  const lockTimeBuf = Buffer.alloc(4, 0); // Locktime 0

  // COINB2 = Sequence + Output Count + Outputs + Locktime
  const coinb2Buf = Buffer.concat([
    sequenceBuf,
    outputCountBuf,
    minerOutputBuf,
    feeOutputBuf,
    witnessOutputBuf,
    lockTimeBuf,
  ]);

  return {
    coinb1: coinb1Buf.toString('hex'),
    coinb2: coinb2Buf.toString('hex'),
  };
}
